docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst

```diff
@@ -0,0 +1,421 @@
+
+For almost all situations we only need to instantiate a dataflow configuration and specify our sources and sinks, and dataflow will handle everything for us.
+
+However, when a write to a field is not visible to CodeQL (for example because it happens in a function whose definition is not in the database) we need to track the qualifier, and tell the dataflow library that it is okay to transfer from flow from the qualifier to the field access. This is done by adding an ``isAdditionalFlowStep`` predicate to the dataflow module.
+
+When adding additional flow steps involving pointers one needs to be careful about whether the dataflow step should flow from the pointer or its indirection. Similarly, care must be taken to decide whether the additional step should target a pointer or its indirection.
+
+In the dual situation where a read of a field is not visible to CodeQL, the dataflow library's "implicit read" mechanism can be used to achieve the right flow.
+
+The setup
+---------
+
+Consider the following scenario: We have a two structs, each containing two fields, and we have two types of ways of getting user input, and two places that data can end up in:
+
+.. code-block:: cpp
+
+  struct A {
+    const int *p;
+    int x;
+  };
+
+  struct B {
+    A *a;
+    int y;
+  };
+
+  void sink(int);
+  int user_input();
+  void write_user_input_to(void *);
+  void read_data(const void *);
+
+A regular dataflow query such as the following query
+
+.. code-block:: ql
+
+  /**
+  * @kind path-problem
+  */
+
+  import semmle.code.cpp.dataflow.new.DataFlow
+  import Flow::PathGraph
+
+  module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) {
+      source.asExpr().(Call).getTarget().hasName("user_input")
+    }
+
+    predicate isSink(DataFlow::Node sink) {
+      exists(Call call |
+        call.getTarget().hasName("sink") and
+        sink.asExpr() = call.getAnArgument()
+      )
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+will catch most things á la
+
+.. code-block:: cpp
+
+  void fill_structure(B* b, const int* pu) {
+    // ...
+    b->a->p = pu;
+  }
+
+  void process_structure(const B* b) {
+    sink(*b->a->p);
+  }
+
+  void get_and_process() {
+    int u = user_input();
+    B* b = (B*)malloc(sizeof(B));
+    // ...
+    fill_structure(b, &u);
+    // ...
+    process_structure(b);
+    free(b);
+  }
+
+This is simple to match because we see:
+  1. User input starts at ``user_input()`` and flows into ``fill_structure``.
+  2. The data is written to the object ``b`` with access path ``[a, p]``.
+  3. The object ``b`` flows out of ``fill_structure`` and into ``process_structure``.
+  4. The access path ``[a, p]`` is read in ``process_structure`` and the value ends up in the sink.
```

Comment:
We could definitely do that. I wanted to convey the fact that the "standard thing" doesn't just work for simple things where we store an integer in a field and later read it out, but also handles the fact that we're storing a pointer to something and later dereference it (because that was a misconception we had to explain to Microsoft in the past).

Reply:
I see your point.  Happy with this as it is.